### PR TITLE
Windows: Silent update except on driver upgrades

### PR DESCRIPTION
**Pull request by zanderz**
I tried it, it works.
@gabriel @chrisnojima

### CODE REVIEW

File: vendor/golang.org/x/sys/windows/registry/key.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+//
+//	s, _, err := k.GetStringValue("SystemRoot")
+//	if err != nil {
+//		log.Fatal(err)
+//	}
+//	fmt.Printf("Windows system root is %q\n", s)
+//
+package registry
+
+import (
```
Comment (chrisnojima) [inline vendor/golang.org/x/sys/windows/registry/key.go]: there should be unit tests for this thing

File: keybase/platform_windows.go [hunk 2]
```diff
@@ -62,6 +66,63 @@ func (c context) BeforeUpdatePrompt(update updater.Update, options updater.Updat
 	return nil
 }
 
+type regUninstallGetter func(string, Log) bool
+
+// It turns out a Wix bundle .exe supports "/layout", which among other things
+// generates a log of what it would do. We can parse this for Dokan product code variables,
+// which will be in the registry if the same version is already present.
+func CheckCanBeSilent(path string, log Log, regFunc regUninstallGetter) bool {
```
Comment (gabriel) [inline keybase/platform_windows.go]: return (bool, error)

File: keybase/platform_windows.go [hunk 3]
```diff
@@ -62,6 +66,63 @@ func (c context) BeforeUpdatePrompt(update updater.Update, options updater.Updat
...
+	tempName, err := util.WriteTempFile("keybaseInstallLayout", []byte{}, 0700)
```
Comment (gabriel) [inline keybase/platform_windows.go]: maybe call the file: "KeybaseInstallLayout.log"?

File: keybase/platform_windows.go [hunk 4]
```diff
@@ -62,6 +66,63 @@ func (c context) BeforeUpdatePrompt(update updater.Update, options updater.Updat
...
+	if err != nil {
+		log.Errorf("CheckCanBeSilent: Unable to write temp file")
+		return false
+	}
+	defer util.RemoveFileAtPath(tempName)
+
+	command := exec.Command(path, "/layout", "/quiet", "/log", tempName)
```
Comment (gabriel) [inline keybase/platform_windows.go]: use command.Exec with timeout

File: keybase/platform_windows.go [hunk 5]
```diff
@@ -62,6 +66,63 @@ func (c context) BeforeUpdatePrompt(update updater.Update, options updater.Updat
...
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	codeFound := false
+	// Of the form: "Variable: DokanProduct64 = {65A3A964-3DC3-0100-0000-160621082245}"m
+	re := regexp.MustCompile(`Variable: DokanProduct(64|86) = (\{([[:xdigit:]]|-)+\})`)
+	for scanner.Scan() {
+		// Give ourselves a way to override silent install in the future
+		if strings.Contains(scanner.Text(), "Variable: KeybaseForceUI") {
+			log.Info("CheckCanBeSilent: KeybaseForceUI")
```
Comment (gabriel) [inline keybase/platform_windows.go]: This seems like Debug. And I always like to make my log messages read like normal writing.

E.g. log.Debug("Found KeybaseForceUI variable")

File: keybase/platform_windows.go [hunk 6]
```diff
@@ -62,6 +66,63 @@ func (c context) BeforeUpdatePrompt(update updater.Update, options updater.Updat
...
+			return false
+		} else if !codeFound {
+			matches := re.FindStringSubmatch(scanner.Text())
+			if len(matches) > 2 {
+				codeFound = regFunc(matches[2], log)
```
Comment (gabriel) [inline keybase/platform_windows.go]: do you want to break if codeFound is true?

File: keybase/platform_windows_test.go [hunk 7]
```diff
@@ -63,3 +63,31 @@ func TestApplyAsset(t *testing.T) {
...
+
+func checkLegit64Code(productID string, log Log) bool {
+	if productID == "{65A3A964-3DC3-0100-0000-160621082245}" {
+		return true
+	}
+	return false
+}
+
+func checkLegit86Code(productID string, log Log) bool {
+	if productID == "{65A3A986-3DC3-0100-0000-160621082245}" {
```
Comment (gabriel) [inline keybase/platform_windows_test.go]: `return productID == "{65A3A986-3DC3-0100-0000-160621082245}"`

and these don't all need to be separate methods..
and make the long strings `"{65A3A986-3DC3-0100-0000-160621082245}"` consts?

File: windows/quietcheck.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+// +build windows
+
+package main
```
Comment (gabriel) [inline windows/quietcheck.go]: can you put this in a subdir  `/windows/quietcheck.go`

File: keybase/platform_windows.go [hunk 9]
```diff
@@ -62,6 +66,63 @@ func (c context) BeforeUpdatePrompt(update updater.Update, options updater.Updat
...
+	}
+	log.Infof("CheckCanBeSilent: returning %v", codeFound)
+	return codeFound
+}
+
+func CheckRegistryUninstallCode(productID string, log Log) bool {
+	log.Infof("CheckCanBeSilent: Searching registry for %s", productID)
+	k, err := registry.OpenKey(registry.LOCAL_MACHINE, `SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\`+productID, registry.QUERY_VALUE|registry.WOW64_64KEY)
+	if err == nil {
+		defer k.Close()
```
Comment (gabriel) [inline keybase/platform_windows.go]: `defer util.Close(k)` and put it right after OpenKey

File: keybase/platform_windows.go [hunk 10]
```diff
@@ -62,6 +66,63 @@ func (c context) BeforeUpdatePrompt(update updater.Update, options updater.Updat
...
+			if len(matches) > 2 {
+				codeFound = regFunc(matches[2], log)
+			}
+		}
+	}
+	log.Infof("CheckCanBeSilent: returning %v", codeFound)
+	return codeFound
+}
+
+func CheckRegistryUninstallCode(productID string, log Log) bool {
```
Comment (gabriel) [inline keybase/platform_windows.go]: make all these methods private

File: vendor/golang.org/x/sys/windows/registry/key.go (hunk 1 above)
Comment (gabriel) [inline vendor/golang.org/x/sys/windows/registry/key.go]: govendor doesn't install tests

File: keybase/platform_windows.go (hunk 3 above)
Comment (gabriel) [inline keybase/platform_windows.go]: oh this is prefix... 

I've been meaning to have file extension support on that method... but for now you should put a dash or period... like `KeybaseInstallLayout.` or `KeybaseInstallLayout-` for the prefix.

path := TempPath("",)

File: keybase/platform_windows.go [hunk 11]
```diff
@@ -62,6 +65,66 @@ func (c context) BeforeUpdatePrompt(update updater.Update, options updater.Updat
...
+
+// CheckCanBeSilent - Interrogate the incoming installer for its driver's
+// uninstall codes. It turns out a Wix bundle .exe supports "/layout", which
+// among other things generates a log of what it would do. We can parse this
+// for Dokan product code variables, which will be in the registry if the same
+// version is already present.
+func CheckCanBeSilent(path string, log Log, regFunc regUninstallGetter) (bool, error) {
+	tempName := util.TempPath("", "keybaseInstallLayout-")
+
+	_, err := command.Exec(path, []string{"/layout", "/quiet", "/log", tempName}, 10*time.Second, log)
```
Comment (gabriel) [inline keybase/platform_windows.go]: I'd make the timeout longer for slow computers... maybe a minute or 2

File: keybase/platform_windows_test.go [hunk 12]
```diff
@@ -63,3 +63,20 @@ func TestApplyAsset(t *testing.T) {
 	err = ctx.Apply(update, updater.UpdateOptions{}, tmpDir)
 	require.NoError(t, err)
 }
+
+const legit64Code = "{65A3A964-3DC3-0100-0000-160621082245}"
+const legit86Code = "{65A3A986-3DC3-0100-0000-160621082245}"
+const mismatchedCode = "{65A3A986-3DC3-0100-0000-160621082244}"
+
+func callCheckCanBeSilent(t *testing.T, path string, code string) bool {
```
Comment (gabriel) [inline keybase/platform_windows_test.go]: I'd name this method `testCheckCanBeSlient`

**Comment by gabriel** [comment]
👍

**Pull request merged**